### PR TITLE
Feature/jenkins 36211

### PR DESCRIPTION
**Pull request by scherler**
Related to issue # . jenkins 36211

Summary of this pull request: 

- [X] Implement logic to stop follow along with key_up and scroll
- [X] Change onClick of nodes when they are running, will now redirect to general url and not set nodeId
- [X] Add unit tests for store and logic component
- [X] add blueocean-acceptance-test for test coverage -> https://github.com/jenkinsci/blueocean-acceptance-test/pull/3

.__Demo videos__ ...

[blueocean-acceptance-test](https://youtu.be/ZOs_gVG6D7E)

@reviewbybees

### CODE REVIEW

File: blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx [hunk 1]
```diff
@@ -15,71 +17,140 @@ import {
...
+                this.setState({ followAlong: false });
             }
-        });
+        };
+        // determine scroll area
+        const domNode = ReactDOM.findDOMNode(this.refs.scrollArea);
+        // add both listemer, one to the scroll area and another to the whole document
+        domNode.addEventListener('wheel', onScrollHandler, false);
+        document.addEventListener('keydown', _handleKeys, false);
     }
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: Should these listeners be cleaned up in `componentWillUnmount()` ?

File: blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx [hunk 2]
```diff
@@ -15,71 +17,140 @@ import {
...
+                // we may have a streaming log
                 const newStart = log.newStart;
                 if (Number(newStart) > 0) {
-                    // kill current  timeout if any
-                    clearTimeout(this.timeout);
-                    this.timeout = setTimeout(() => fetchLog({ ...logGeneral, newStart }), 1000);
+                    // in case we doing karaoke we want to see more logs
+                    if (this.state.followAlong) {
+                        // kill current  timeout if any
+                        clearTimeout(this.timeout);
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: We may also need to clear this timeout in unmount?

File: blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx [hunk 3]
```diff
@@ -15,71 +17,140 @@ import {
...
+        const nodeAction = calculateNode(this.props, nextProps, this.mergedConfig);
+        if (nodeAction && nodeAction.action) {
+            // use updated config
+            this.mergedConfig = nodeAction.config;
+            // we may need to stop following
+            if (this.state.followAlong !== nodeAction.state.followAlong) {
+                this.setState({ followAlong: nodeAction.state.followAlong });
+            }
+            // if we have actions we fire them
+            this.props[nodeAction.action](this.mergedConfig);
```
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: This dynamic invoke might be a little hard for some people to follow. I am not strongly against it, but I wonder if there's a way to do this in a way that's a little more obvious.

**Comment by cliffmeyers** [comment]
LGTM pending the two comments above. I think we need to make sure the event listeners aren't leaking before we bee it.

File: blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx (hunk 3 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: I extracted it mainly to make it testable. I was playing here with a way to calculate followUps which are completly covered by unit tests, so testing becomes better.

File: blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx (hunk 1 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: fixed that, thanks for spotting.

**Comment by michaelneale** [comment]
Tested it out, some small 🐜's: 

1) When I launch a job (multibranch) and open it - it jumps to the "final" stage, even though it hasn't really started yet - which seems odd (haven't noticed it)
2) I am unable to run a branch name of "feature/newhotness" 

To reproduce: 

I created the https://github.com/cloudbeers/blueocean-demo-project project. Add it as a multibranch project, and then go to the "branches" tab - press run on master, or run on the branch, to see the above to problems. 

```
jenkins-js-extension.js:10841 POST http://localhost:8080/jenkins/blue/rest/organizations/jenkins/pipelines/multi2/branches/feature%2Fnewhotness/runs/ 404 (Not Found)run @ jenkins-js-extension.js:10841run @ jenkins-js-extension.js:13874onClick @ jenkins-js-extension.js:13938ReactErrorUtils.invokeGuardedCallback @ blueocean.js:28058executeDispatch @ blueocean.js:20044executeDispatchesInOrder @ blueocean.js:20064executeDispatchesAndRelease @ blueocean.js:19517executeDispatchesAndReleaseTopLevel @ blueocean.js:19528forEachAccumulated @ blueocean.js:34385processEventQueue @ blueocean.js:19690runEventQueueInBatch @ blueocean.js:28084handleTopLevel @ blueocean.js:28095handleTopLevelImpl @ blueocean.js:28173perform @ blueocean.js:33838batchedUpdates @ blueocean.js:27265batchedUpdates @ blueocean.js:31544dispatchEvent @ blueocean.js:28250
jenkins-js-extension.js:13881 Unexpected error queuing a run of "feature%2Fnewhotness". Response:(anonymous function) @ jenkins-js-extension.js:13881(anonymous function) @ jenkins-js-extension.js:10851
jenkins-js-extension.js:13882 Response {type: "basic", url: "http://localhost:8080/jenkins/blue/rest/organizati…elines/multi2/branches/feature%2Fnewhotness/runs/", status: 404, ok: false, statusText: "Not Found"…}
```

**Comment by scherler** [comment]
@sophistifunk regarding the 2 issue seems not related to the PR at first sight, but will try now. First issue I am trying to reproduce.

**Comment by michaelneale** [comment]
@scherler yes sorry I didnt' validate it was not this PR, I suspect it may not be, but ran out of time to check with other things. Looking good!

**Comment by michaelneale** [comment]
ok, ignore #2 error.

**Comment by michaelneale** [comment]
OK LGTM based on @cliffmeyers's comments, and perhaps the issue number 1 (if not resolved, then noted and raised). Issue number 2 (the clicking on feature/newhotness) is NOT related to this PR, so you can ignore that.

**Comment by cliffmeyers** [comment]
LGTM Thor 🐝

**Pull request merged**